beacon-chain/types/block.go

```diff
@@ -1,72 +1,120 @@
+		h, err := blake2b.New256(data.CrystallizedStateHash)
+		if err != nil {
+			return nil, err
+		}
+		activeStateHash = h
+	}
+	return &Block{data, activeStateHash, crystallizedStateHash}, nil
+}
+
+// NewGenesisBlock returns the canonical, genesis block for the beacon chain protocol.
+func NewGenesisBlock() (*Block, error) {
+	genesisTime := time.Date(2018, time.July, 21, 12, 0, 0, 0, time.UTC)
+	protoGenesis, err := ptypes.TimestampProto(genesisTime)
 	if err != nil {
 		return nil, err
 	}
+	// TODO: Add more default fields.
+	return &Block{data: &pb.BeaconBlockResponse{Timestamp: protoGenesis}}, nil
+}
+
+// Hash generates the blake2b hash of the block
+func (b *Block) Hash() (hash.Hash, error) {
+	data, err := proto.Marshal(b.data)
+	if err != nil {
+		return nil, fmt.Errorf("could not marshal block proto data: %v", err)
+	}
 	return blake2b.New256(data)
 }
 
-// Data getter makes the block's properties read-only.
-func (b *Block) Data() *Data {
-	return b.data
+// ParentHash corresponding to parent beacon block.
+func (b *Block) ParentHash() (hash.Hash, error) {
+	return blake2b.New256(b.data.ParentHash)
 }
 
-// NewBlock creates a new beacon block given certain arguments.
-func NewBlock(slotNumber uint64) *Block {
-	data := &Data{Timestamp: time.Now(), SlotNumber: slotNumber}
-	return &Block{data}
+// SlotNumber of the beacon block.
+func (b *Block) SlotNumber() uint64 {
+	return b.data.SlotNumber
 }
 
-// NewGenesisBlock returns the canonical, genesis block for the beacon chain protocol.
-func NewGenesisBlock() *Block {
-	timestamp := time.Date(2018, time.July, 21, 12, 0, 0, 0, time.UTC)
-	// TODO: Add more default fields.
-	return &Block{data: &Data{Timestamp: timestamp}}
+// MainChainRef returns a keccak256 hash corresponding to a PoW chain block.
+func (b *Block) MainChainRef() common.Hash {
+	return common.BytesToHash(b.data.MainChainRef)
 }
 
-// Data contains the fields in a beacon chain block.
-type Data struct {
-	ParentHash              hash.Hash       // ParentHash is the hash of the parent beacon block.
-	SlotNumber              uint64          // Slot number is the number a client should check to know when it creates block.
-	RandaoReveal            hash.Hash       // RandaoReveal is used for Randao commitment reveal.
-	AttestationBitmask      []byte          // AttestationBitmask is the bit field of who from the attestation committee participated.
-	AttestationAggregateSig []uint          // AttestationAggregateSig is validator's aggregate sig.
-	ShardAggregateVotes     []AggregateVote // ShardAggregateVotes is shard aggregate votes.
-	MainChainRef            common.Hash     // MainChainRef is the reference to main chain block.
-	ActiveStateHash         hash.Hash       // ActiveStateHash is the state that changes every block.
-	CrystallizedStateHash   hash.Hash       // CrystallizedStateHash is the state that changes every epoch.
-	Timestamp               time.Time
+// RandaoReveal returns the blake2b randao hash.
```

Comment:
Can you please add a comment this is just temporary?  Randao reveal happens after proposer signs a block which generates a random number